### PR TITLE
Fix crash when setting value for relation widget wrapper

### PR DESCRIPTION
**Pull request by nyalldawson**
Calling setForeignKeys during updateValues would result in the
widget wrapper catching the relation widget's changed signal,
triggering an assert as the foreign keys don't match the form's
feature's key (since that hasn't yet been set).

In any case, setting the initial value for a editor widget wrapper
must NOT be treated as a user-induced value change.

### CODE REVIEW

**Comment by nyalldawson** [comment]
Can anyone review this prior to release? It's a crash fix for a regression caused by newly introduced functionality in a widely used post of qgis...

**Comment by roya0045** [comment]
using a ++ -- to toggle a behavior seems hacky. In other places there are dedicated connect and disconnect to prevent triggering. But if it works, I guess its fine.

**Comment by nyalldawson** [comment]
@roya0045 I agree it's not typical, but I've started doing this approach for a few reasons:

- the standard blockSignals approach is a nuclear bomb approach which prevents ALL signals from the object, regardless of what they are and what is connected to them. It should only be used in very limited circumstances, because even internally a widget may have connections with itself to change different parts of its behaviour as values change.
- a bool is not a great solution, because if there's any chance that another function is also using that same variable we run the risk of nested calls causing it to clear the value earlier then we want.
- using an int is nice and safe, because it means that we wait for ALL code which wants to block the updates to finish before we unlock them again, and there's zero risk of nesting issues playing havoc with the logic 😁

**Comment by nyalldawson** [comment]
Oh, and I consider dedicated disconnect/connect blocks to also be dangerous due to the previously mentioned issues with nested calls (same issue as using bools)

**Comment by roya0045** [comment]
This is what I had in mindhttps://github.com/qgis/QGIS/blob/f32725658f205af4ac9718d3a3b74f13eaa98d3e/src/gui/layout/qgslayoutscalebarwidget.cpp#L178

along with the related connect function.

As it disable a connection temporarily without blocking signals. I reimplemented that for the layoutlabel but I didn't seem to have a positive affect in that case.

**Review by roya0045 — APPROVED** [review]
Seems safe enough.

**Comment by nyalldawson** [comment]
FWIW, I struggled for ages to write a meaningful test for this and failed

File: src/gui/editorwidgets/qgsrelationreferencewidgetwrapper.cpp [hunk 1]
```diff
@@ -218,6 +220,9 @@ void QgsRelationReferenceWidgetWrapper::setEnabled( bool enabled )
 
 void QgsRelationReferenceWidgetWrapper::foreignKeysChanged( const QVariantList &values )
 {
+  if ( mBlockChanges ) // initial value is being set, we can ignore this signal
```
Comment (elpaso) [inline src/gui/editorwidgets/qgsrelationreferencewidgetwrapper.cpp:223]:
```suggestion
  if ( mBlockChanges != 0  ) // initial value is being set, we can ignore this signal
```

just for clarity to remind the reader the it isn't a bool

**Review by elpaso — APPROVED** [review]

File: src/gui/editorwidgets/qgsrelationreferencewidgetwrapper.cpp [hunk 2]
```diff
@@ -218,6 +220,9 @@ void QgsRelationReferenceWidgetWrapper::setEnabled( bool enabled )
 
 void QgsRelationReferenceWidgetWrapper::foreignKeysChanged( const QVariantList &values )
 {
+  if ( mBlockChanges != 0  ) // initial value is being set, we can ignore this signal
```
Comment (nyalldawson) [inline src/gui/editorwidgets/qgsrelationreferencewidgetwrapper.cpp:223]:
```suggestion
  if ( mBlockChanges != 0 ) // initial value is being set, we can ignore this signal
```

**Pull request merged**